CONFIGURATION.md
# Metalnx Deployment Configuration

Metalnx has a bias towards deployment as a Docker image, and will continue to evolve so that it can be clustered
and scaled horizontally using an orchestration tool such as Kubernetes, as
part of a suite of mid-tier components that can work together to create a complete CI environment.

## /etc/irods-ext

This project contains a sample `/etc/irods-ext` directory with a `metalnx.properties` and `metalnxConfig.xml` examples.
These two files must be mounted as a volume at `/etc/irods-ext`, or be present in the same directory when running
in a web container directly on the host OS.

Generally, Metalnx is configured through a `metalnx.properties` file, documented below. This controls the optional behaviors
and customizations with the exception of theming (custom images, banners, css, messages). 

In addition, theming via assets is controlled by the required `metalnxConfig.xml` file. This file must be in `/etc/irods-ext` or
you will receive invalid resource messages. The provided file will cause the default theme and messages to be utilized.

### App configuration (metalnx.properties)

`metalnx.properties` is read from `/etc/irods-ext/metalnx.properties` during deployment within the container. This file controls the optional behaviors and customizations. This file must be present for Metalnx to run.

### App theming (metalnxConfig.xml)

By default, Metalnx uses the resource support of [Spring MVC](https://docs.spring.io/spring/docs/4.1.9.RELEASE/spring-framework-reference/html/mvc.html#mvc-config-static-resources).

The provided `/etc/irods-ext/metalnxConfig.xml` can be used to load the default baseline theming. This file must be present for Metalnx to run.  It controls the theming of the deployed application, including custom images, banners, css, javascript, and any internationalization of the user messages.

## Running Docker with these configs

A local directory with these two files must be volume mounted when running via Docker.

An example developer run of the docker image with custom configuration:

```
docker run -it \
-p 8080:8080 \
-v `pwd`/mylocal-irods-ext:/etc/irods-ext \
mylocal/metalnx:dev
```


docs/PAM-&-SSL-Configuration.md
This document explains how to configure Metalnx to work with an iRODS grid set up to use SSL and PAM.

There are few different ways to use SSL with or without PAM in your system:

![](https://github.com/Metalnx/metalnx-web/blob/master/docs/IMAGES/mlx-ssl-config.png)

## iRODS

In this section, we provide a quick guide on how to configure iRODS with SSL. If you want more information about it, we encourage you to check out this [SSL and PAM presentation](http://slides.com/irods/ugm2018-ssl-and-pam-configuration#/) by iRODS.

To configure SSL in iRODS you need to create a directory on the catalog provider to hold the certificates. For the purpose of this document, we use self-signed certificates, but the process is similar if you have a certificate issued by a certificate authority. 

Run the following commands to create the private key, certificate, and Diffie-Hellman parameters in `/etc/irods/ssl/`:
        
    $ mkdir /etc/irods/ssl
    $ cd /etc/irods/ssl
    $ openssl genrsa -out irods.key                                     # Generate an RSA key
    $ openssl req -new -x509 -key irods.key -out irods.crt -days 365    # Generate self-signed certificate
    $ openssl dhparam -2 -out dhparams.pem 2048                         # Generate some Diffie-Hellman parameters
    $ chmod 600 server.key                                              # Limit permission on the key

Make sure your */etc/irods/ssl* looks like:

```
$ ls -l /etc/irods/ssl
-rw-rw-r-- 1 irods irods 1277 Dec 13 01:48 irods.crt
-rw-rw-r-- 1 irods irods  424 Dec 13 01:50 dhparams.pem
-rw------- 1 irods irods 1675 Dec 13 01:46 server.key
```

In addition, change the catalog provider to require SSL for all connections.  Update the `/etc/irods/core.re` file to use `CS_NEG_REQUIRE` as follows:

    acPreConnect(*OUT) { *OUT="CS_NEG_REQUIRE"; }

Now, update the clients to use SSL. The first client we will update is the iRODS service account. Your `irods_environment.json` shoud be similar to this:

```
{
    "irods_host": "provider.example.org",
    "irods_port": 1247,
    "irods_default_resource": "demoResc",
    "irods_home": "/tempZone/home/rods",
    "irods_cwd": "/tempZone/home/rods",
    "irods_user_name": "rods",
    "irods_zone_name": "tempZone",
    "irods_client_server_negotiation": "request_server_negotiation",
    "irods_client_server_policy": "CS_NEG_REQUIRE",
    "irods_encryption_key_size": 32,
    "irods_encryption_salt_size": 8,
    "irods_encryption_num_hash_rounds": 16,
    "irods_encryption_algorithm": "AES-256-CBC",
    "irods_default_hash_scheme": "SHA256",
    "irods_match_hash_policy": "compatible",
    "irods_server_control_plane_port": 1248,
    "irods_server_control_plane_key": "TEMPORARY__32byte_ctrl_plane_key",
    "irods_server_control_plane_encryption_num_hash_rounds": 16,
    "irods_server_control_plane_encryption_algorithm": "AES-256-CBC",
    "irods_maximum_size_for_single_buffer_in_megabytes": 32,
    "irods_default_number_of_transfer_threads": 4,
    "irods_transfer_buffer_size_for_parallel_transfer_in_megabytes": 4,
    "irods_ssl_certificate_chain_file": "/etc/irods/ssl/irods.crt",
    "irods_ssl_certificate_key_file": "/etc/irods/ssl/server.key",
    "irods_ssl_dh_params_file": "/etc/irods/ssl/dhparams.pem",
    "irods_ssl_ca_certificate_file": "/etc/irods/ssl/irods.crt",
    "irods_ssl_verify_server": "cert"
}
```
Notice that we changed the `irods_client_server_policy` to `CS_NEG_REQUIRE` and added SSL configuration parameters:

    "irods_client_server_policy": "CS_NEG_REQUIRE",
    "irods_ssl_certificate_chain_file": "/etc/irods/ssl/irods.crt",
    "irods_ssl_certificate_key_file": "/etc/irods/ssl/server.key",
    "irods_ssl_dh_params_file": "/etc/irods/ssl/dhparams.pem",
    "irods_ssl_ca_certificate_file": "/etc/irods/ssl/irods.crt"

Restart the iRODS service:

    ./irodsctl restart

### Log in iRODS

After all that configuration, you should be able to log in iRODS without any problems.

	$ iinit
	Enter your current iRODS password:
	$ ils
	/tempZone/home/rods

### PAM Configuration

This section provides you a quick explanation on how to set up PAM authentication in iRODS. If you do not want use PAM in your system, go to the Metalnx Configuration section.

### iRODS PAM file

First, we need to configure the iRODS PAM file:

    sudo su - root -c 'echo "auth sufficient pam_unix.so" > /etc/pam.d/irods'

In this case, the `pam_unix.so` module is used for traditional password authentication.

### Unix User creation

Next, if you want to use the `pam_unix.so` module, every iRODS user must be a Unix user. So, before adding anybody to iRODS make sure you create the corresponding user in unix. For example, with the following lines, we create the user *bob* and set his password by running:

    $ sudo useradd bob
    $ sudo passwd bob

*bob* will be the user used to authenticate in iRODS via PAM.

### iRODS User Creation

This *bob* user still needs to be added to the iRODS database (no need to set a password in iRODS). Do that by running:

    $ iadmin mkuser bob rodsuser

The following instructions are only necessary **if you want to use a command line to interact with iRODS. Metalnx does not require any of the following changes**.

Modify the `irods_environment.json` file on the client to use PAM as the authentication scheme:

    "irods_authentication_scheme": "PAM"

iRODS comes with a tool used for testing basic PAM authentication. You can run it to check if you are able to authenticate:

    iRODS/server/bin/PamAuthCheck metalnx

*Note that this program waits for you to type the password without any prompt. So, after running the command above ensure that you type your password correctly. This tool returns `Authenticated` or `Not Authenticated`*.

Now, update the `irods_environment.json` file on the client, set the `irods_user_name` to *bob* and the `irods_authentication_scheme` to PAM:

```
{
    "irods_host": "provider.example.org",
    "irods_zone_name": "tempZone",
    "irods_port": 1247,
    "irods_user_name": "bob",
    "irods_authentication_scheme": "PAM",
    "irods_client_server_negotiation": "request_server_negotiation",
    "irods_client_server_policy": "CS_NEG_REQUIRE",
    "irods_ssl_ca_certificate_file": "/etc/irods/ssl/irods.crt",
    "irods_encryption_key_size": 16,
    "irods_encryption_salt_size": 8,
    "irods_encryption_num_hash_rounds": 16,
    "irods_encryption_algorithm": "AES-256-CBC"
}
```

## Metalnx Configuration

The next step is to configure the Metalnx application. Before going any further, make sure you have Java 8 or higher installed. You can check your current java version by running:

    $ java -version
    java version "1.8.0_121"
    Java(TM) SE Runtime Environment (build 1.8.0_121-b13)
    Java HotSpot(TM) 64-Bit Server VM (build 25.121-b13, mixed mode)

Now, we need to tell the Java Virtual Machine to trust the irods certificate created earlier. This can be done by running the following command:

    cd /etc/irods/ssl
    keytool -import -alias irodscertificate -file /etc/irods/ssl/irods.crt -keystore irodskeystore

The command above will create a keystore called *irodskeystore* in `/etc/irods/ssl`. This keystore is necessary for the `setup_script.py` to check whether or not Metalnx can securely connect to iRODS.

After creating the irods keystore file, you can run `python /opt/emc/setup_script.py` to configure Metalnx. At this point, you do not have to provide any additional information about keystores, the script takes care of the setup automatically.

Finally, there is only one configuration to be done. Specify to Tomcat where the `irodskeystore` file is, so the Metalnx Web server is able to communicate with iRODS using SSL. To do so, modify the `JAVA_OPTS` option in the file `/etc/default/tomcat7`:

    JAVA_OPTS="-Djavax.net.ssl.trustStore=/etc/irods/ssl/irodskeystore -Djavax.net.ssl.trustStorePassword=<keystore-password>"

### HTTPS

If you want to use HTTPS protocol for secure communication, you can set it up with the Metalnx setup script. 

During the setup process (`python /opt/emc/setup_script.py`), one of the last steps is to configure HTTPS. The script asks following question:

    Do you want to enable HTTPS on Tomcat and Metalnx? [Y/n]:  

If you answer ***n*** Tomcat will respond to the HTTP protocol on port 8080. If you press ***Y*** or ***y***, it will configure Tomcat to use SSL encryption (HTTPS) on port 8443.

If you encounter a message in the script that reads:

    Starting Tomcat ... Failed. Tomcat could not be started. Please do it manually.

This message means that the installation script could not restart Tomcat on its own.  You will need to take this action manually using the systemctl command (see below).

     # systemctl start tomcat


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

### Added

- [#317] Build no longer pollutes personal .m2 directory.
- [#191] Added support for hiding metadata based on a set of prefixes.

### Changed

- [#247,#337,#343] Bumped version of Jargon for bug fixes.
- [#294] Create IRODSFileSystem and IRODSAccessObjectFactory only once.
- [#332] Set createParentCollections to true when calling createCollection
- [#191] Made it so that rodsadmins can always see metadata.
- [#321] Bumped Jargon version.
- [#255] User is now asked to confirm changes to permissions.


packaging/scripts/lib/config.py
#	Copyright (c) 2015-2017, Dell EMC.
#
#	Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
import getpass
from base64 import b64encode
from hashlib import md5
from socket import gethostname
from os import path, getenv

KEY_VAL_SEPARATOR = '='

SALT = '!M3t4Lnx@1234'

SETUP_FILE = '.env'


def read_env_file(file_path):
    """
    Function that parses a .env file and return a key-value dictionary
    :param file_path:
    :return: dictionary
    """
    if not file_path or not path.isfile(file_path):
        return None

    data = {}

    with open(file_path, 'r') as f:
        for line in f:
            c = line.split(KEY_VAL_SEPARATOR)
            if len(c) < 2:
                return
            else:
                data[c[0]] = c[1].strip()

    return data


class MetalnxConfigEnv(object):
    def __init__(self, source):
        self.data = read_env_file(source)
        self.data_cache = {}

    def get(self, key):
        return self.data[key] if self.data and key in self.data else ''

    def set(self, option, value):
        """
        Saves all key, values params into memory to be dumped into a file later
        :param option: config param name
        :param value: value for config param
        :return: None
        """
        if not option or not value:
            return

        self.data_cache[option] = value

    def dump(self):
        """
        Dumps new configuration params into the setup file
        :return: None
        """
        if not self.data_cache:
            return

        # keeping existing params that were not set this time
        if self.data:
            for k, v in self.data.iteritems():
                if k not in self.data_cache and v != '':
                    self.data_cache[k] = v

        # writing all params to file
        with open(SETUP_FILE, 'w+') as f:
            for k, v in self.data_cache.iteritems():
                param = '{}={}\n'.format(k, v)
                f.write(param)


class MetalnxConfig(object):

    def __init__(self):
        self.setup = MetalnxConfigEnv(SETUP_FILE)

    def __call__(self, *args, **kwargs):
        """
        Shortcut to get
        """
        return self.get(*args, **kwargs)

    def get(self, option, default=None):
        if not self.setup:
            return None

        value = self.setup.get(option) or default
        return value if value is not None else ''

    def set(self, option, value):
        """
        Sets a new value for an option
        :param option: config param to be kept
        :param value: config param value to be kept
        :return: None
        """
        self.setup.set(option, value)

    def save(self):
        """
        Saves all config params to .env file
        :return:
        """
        self.setup.dump()

config = MetalnxConfig()


def encode_password(pwd):
    """Encodes password"""

    def pick_key(hostname):
        hostname = hostname.split('.')[0]
        s = md5(u'{}{}'.format(SALT, hostname)).digest()
        return sum([ord(c) for c in s])

    def encode(s, k):
        return ''.join([chr((ord(c) ^ k) % 256) for c in s])

    return b64encode(encode(pwd, pick_key(gethostname())))


def get_mlx_url(is_https):
    if is_https:
        return MLX_URL_FORMAT('https', gethostname(), '8443')
    return MLX_URL_FORMAT('http', gethostname(), '8080')

RELEASE_VERSION = '1.1.0'
BUILD_NUMBER = ''

TEST_CONNECTION_JAR = '/opt/emc/test-connection.jar'

KEYSTORE_DIR = '/etc/irods/ssl'
KEYSTORE_FILE = 'irodskeystore'
JAVA_SSL_TRUSTSTORE_PARAM = '-Djavax.net.ssl.trustStore'

IRODS_PROPS_FILENAME = 'irods.environment.properties'
DATABASE_PROPS_FILENAME = 'database.properties'

POSTGRESQL = 'postgresql'
MYSQL = 'mysql'

GITHUB_URL = 'https://github.com/Metalnx/metalnx-web'
MLX_URL_FORMAT = '{}://{}:{}/emc-metalnx-web/login/'.format

HIBERNATE_CONFIG = {
    MYSQL: {
        'db.driverClassName': 'com.mysql.jdbc.Driver',
        'hibernate.dialect': 'org.hibernate.dialect.MySQL5Dialect',
    },
    POSTGRESQL: {
        'db.driverClassName': 'org.postgresql.Driver',
        'hibernate.dialect': 'org.hibernate.dialect.PostgreSQLDialect'
    },
    'options': {
        'hibernate.show_sql': 'false',
        'hibernate.format_sql': 'false',
        'hibernate.hbm2ddl.auto': 'update',
        'connection.pool_size': '10'
    }
}

IRODS_PROPS_SPEC = {
    'host': {
        'name': 'irods.host',
        'desc': 'iRODS Host',
        'env_param': 'IRODS_HOST',
        'default': lambda c: config('IRODS_HOST', default='localhost'),
        'order': 1
    },
    'port': {
        'name': 'irods.port',
        'desc': 'iRODS Port',
        'env_param': 'IRODS_PORT',
        'default': lambda c: config('IRODS_PORT', default='1247'),
        'order': 2
    },
    'zone': {
        'name': 'irods.zoneName',
        'desc': 'iRODS Zone',
        'env_param': 'IRODS_ZONE',
        'default': lambda c: config('IRODS_ZONE', default='tempZone'),
        'order': 3
    },
    'auth_scheme': {
        'name': 'irods.auth.scheme',
        'values': ['STANDARD', 'GSI', 'PAM', 'KERBEROS'],
        'desc': 'Authentication Schema',
        'env_param': 'IRODS_AUTH_SCHEME',
        'default': lambda c: config('IRODS_AUTH_SCHEME', default='STANDARD'),
        'order': 4
    },
    'user': {
        'name': 'jobs.irods.username',
        'desc': 'iRODS Admin User',
        'env_param': 'IRODS_USER',
        'default': lambda c: config('IRODS_USER', default='rods'),
        'order': 5
    },
    'password': {
        'name': 'jobs.irods.password',
        'desc': 'iRODS Admin Password',
        'env_param': None,
        'default': lambda c: '',
        'input_method': getpass.getpass,
        'cast': encode_password,
        'order': 6
    }
}

DB_TYPE_SPEC = {
    MYSQL: {
        'db.driverClassName': 'com.mysql.jdbc.Driver',
        'hibernate.dialect': 'org.hibernate.dialect.MySQL5Dialect'
    },
    POSTGRESQL: {
        'db.driverClassName': 'org.postgresql.Driver',
        'hibernate.dialect': 'org.hibernate.dialect.PostgreSQLDialect'
    },
    'url': {
        'name': 'db.url',
        'cast': 'jdbc:{}://{}:{}/{}'.format
    }
}

DB_PROPS_SPEC = {
    'host': {
        'name': 'db.host',
        'desc': 'Metalnx Database Host',
        'env_param': 'DB_HOST',
        'default': lambda c: config('DB_HOST', default='localhost'),
        'order': 1
    },
    'port': {
        'name': 'db.port',
        'desc': 'Metalnx Database Port',
        'env_param': 'DB_PORT',
        'default': lambda c: config('DB_PORT', default='3306' if c == 'mysql' else '5432'),
        'order': 2
    },
    'db_name': {
        'name': 'db.db_name',
        'desc': 'Metalnx Database Name',
        'env_param': 'DB_NAME',
        'default': lambda c: config('DB_NAME', default='metalnx'),
        'order': 3
    },
    'user': {
        'name': 'db.username',
        'desc': 'Metalnx Database User',
        'env_param': 'DB_USER',
        'default': lambda c: config('DB_USER', default='metalnx'),
        'order': 4
    },
    'password': {
        'name': 'db.password',
        'desc': 'Metalnx Database Password',
        'env_param': None,
        'default': lambda c: '',
        'input_method': getpass.getpass,
        'cast': encode_password,
        'order': 5
    }
}

INSTALL_STEPS = [
    'config_java_devel',
    'config_tomcat_home',
    'config_tomcat_shutdown',
    'config_metalnx_package',
    'config_existing_setup',
    'config_war_file',
    'config_database',
    'config_irods',
    'config_restore_conf',
    'config_set_https',
    'config_confirm_props',
    'config_tomcat_startup',
    'config_displays_summary',
]

tomcat_dirs = {
    'versions': ['', '6', '7', '8', '9'],
    'home': '/usr/share/tomcat',
    'conf': '/etc/tomcat',
    'webapps': '/var/lib/tomcat'
}
